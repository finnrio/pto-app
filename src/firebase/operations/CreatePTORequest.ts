import { addDoc, collection, doc, getDoc, setDoc, updateDoc } from "firebase/firestore";
import { FIRESTORE_DB } from "../firebaseConfig";
import { faker } from "@faker-js/faker";

function calculateHours(startDate: Date, endDate: Date): number {
  const start = new Date(startDate);
  const end = new Date(endDate);
  if (start > end) {
    throw new Error("Ending date must be after starting date");
  }
  let count = 0;
  while (start <= end) {
    const dayOfWeek = start.getDay();
    if (dayOfWeek > 0 && dayOfWeek < 6) {
      count += 1;
    }
    start.setDate(start.getDate() + 1);
  }
  return count * 8; // assumed 8 hours per day
}

export default async function CreatePTORequest(
  userId: string,
  startDate: Date,
  endDate: Date,
  reason: string,
) {
  const { NODE_ENV } = process.env;

  // Calculate the number of hours requested
  const requestedHours = calculateHours(startDate, endDate);

  // get the user's data
  const userDoc = doc(FIRESTORE_DB, `${NODE_ENV}`, userId);
  const loadedDoc = await getDoc(userDoc);
  if (!loadedDoc.exists()) {
    console.log("No user data found");
    throw new Error("No user data found");
  }

  // eslint-disable-next-line camelcase
  const { pto_allowance, pto_pending, pto_used } = loadedDoc.data();

  // check if the user has enough PTO hours available
  // eslint-disable-next-line camelcase
  if (pto_pending + pto_used + requestedHours > pto_allowance) {
    console.log("Not enough PTO hours available");
    throw new Error("Not enough PTO hours available");
  }

  const ptoCollection = collection(userDoc, "pto");

  // Add a new document with a generated id.
  const ptoUuid = faker.string.uuid(); // TODO generate a random ID with UUIDv4library
  await setDoc(doc(ptoCollection, ptoUuid), {
    start_date: startDate.toISOString().split("T")[0],
    end_date: endDate.toISOString().split("T")[0],
    reason,
    status: "Pending",
    user_id: userId,
    hours: requestedHours,
  });

  // Update pto_pending in the user's document and remove from pto_available
  await updateDoc(
    userDoc,
    { pto_pending: pto_pending + requestedHours }, // eslint-disable-line camelcase
  );
  return ptoUuid;
}
