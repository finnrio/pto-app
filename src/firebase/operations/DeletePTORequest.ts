import { deleteDoc, doc, getDoc, updateDoc } from "firebase/firestore";
import { FIRESTORE_DB } from "../firebaseConfig";

export default async function DeletePTORequest(userId: string, ptoId: string) {
  const userDocRef = doc(FIRESTORE_DB, `${process.env.NODE_ENV}`, userId);
  const userData = await getDoc(userDocRef);
  if (!userData.exists()) {
    throw new Error("User not found");
  }

  const ptoDocRef = doc(userDocRef, "pto", ptoId);
  await getDoc(ptoDocRef).then(async (docRef) => {
    if (!docRef.exists()) {
      throw new Error("PTO Request not found");
    } else if (docRef.data().status === "Pending") {
      updateDoc(userDocRef, {
        pto_pending: userData.data().pto_pending - docRef.data().hours,
      });
    } else if (docRef.data().status === "Approved") {
      updateDoc(userDocRef, {
        pto_used: userData.data().pto_used - docRef.data().hours,
      });
    }
    await deleteDoc(
      doc(FIRESTORE_DB, `${process.env.NODE_ENV}`, userId, "pto", ptoId),
    );
  });
}
