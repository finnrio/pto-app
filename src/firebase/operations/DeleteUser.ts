import { deleteDoc, doc } from "firebase/firestore";
import { FIRESTORE_DB } from "../firebaseConfig";

export default async function DeleteUser(uid: string) {
  await deleteDoc(doc(FIRESTORE_DB, `${process.env.NODE_ENV}`, uid));
}
