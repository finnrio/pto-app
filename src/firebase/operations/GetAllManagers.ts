import { collection, query, where, getDocs } from "firebase/firestore";
import { FIRESTORE_DB } from "../firebaseConfig";

export default async function GetAllManagers(): Promise<any[]> {
  const q = query(
    collection(FIRESTORE_DB, `${process.env.NODE_ENV}`),
    where("role", "==", "Manager"),
  );
  const querySnapshot = await getDocs(q);

  const res: any[] = [];

  querySnapshot.forEach((manager) => {
    res.push({
      value: manager.id,
      label: `${manager.data().first_name} ${manager.data().surname}`,
    });
  });
  return res;
}
