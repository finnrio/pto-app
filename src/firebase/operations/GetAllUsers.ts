import { collection, query, where, getDocs } from "firebase/firestore";
import { FIRESTORE_DB } from "../firebaseConfig";

export default async function GetAllUsers(): Promise<any[]> {
  const q = query(
    collection(FIRESTORE_DB, `${process.env.NODE_ENV}`),
    where("role", "==", "User"),
  );
  const querySnapshot = await getDocs(q);

  const res: any[] = [];

  querySnapshot.forEach((user) => {
    res.push({
      value: user.id,
      label: `${user.data().first_name} ${user.data().surname}`
    });
  });
  return res;
}
