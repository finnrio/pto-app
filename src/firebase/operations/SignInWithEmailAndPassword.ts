import { signInWithEmailAndPassword } from "firebase/auth";
import { FIREBASE_AUTH } from "../firebaseConfig";

export default async function SignInWithEmailAndPassword(
  email: string,
  password: string,
): Promise<void> {
  await signInWithEmailAndPassword(FIREBASE_AUTH, email, password)
    .then((userCredentials) => {
      console.log("Signed in successfully as user:", userCredentials.user.uid);
    })
    .catch((error) => {
      console.error(error.code);
      throw error;
    });
}
