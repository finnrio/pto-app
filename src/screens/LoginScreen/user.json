{
    "_redirectEventId": undefined,
    "apiKey": "AIzaSyBEjW_646bH_Q4ECQD_g7nXhM_WTDQQZqw",
    "appName": "[DEFAULT]",
    "createdAt": "1712836751208",
    "displayName": undefined,
    "email": "archie@test.test",
    "emailVerified": false,
    "isAnonymous": false,
    "lastLoginAt": "1712940969539",
    "phoneNumber": undefined,
    "photoURL": undefined,
    "providerData": [[Object]],
    "stsTokenManager": {
        "accessToken": "eyJhbGciOiJSUzI1NiIsImtpZCI6ImYyOThjZDA3NTlkOGNmN2JjZTZhZWNhODExNmU4ZjYzMDlhNDQwMjAiLCJ0eXAiOiJKV1QifQ.eyJpc3MiOiJodHRwczovL3NlY3VyZXRva2VuLmdvb2dsZS5jb20vcHRvLWFwcC0zMDI3MSIsImF1ZCI6InB0by1hcHAtMzAyNzEiLCJhdXRoX3RpbWUiOjE3MTI5NDA5NjksInVzZXJfaWQiOiJYS2dPTUkyZTZVWkJ5ekY5cE04Y2d0NVVTSWIyIiwic3ViIjoiWEtnT01JMmU2VVpCeXpGOXBNOGNndDVVU0liMiIsImlhdCI6MTcxMjk0MDk2OSwiZXhwIjoxNzEyOTQ0NTY5LCJlbWFpbCI6ImFyY2hpZUB0ZXN0LnRlc3QiLCJlbWFpbF92ZXJpZmllZCI6ZmFsc2UsImZpcmViYXNlIjp7ImlkZW50aXRpZXMiOnsiZW1haWwiOlsiYXJjaGllQHRlc3QudGVzdCJdfSwic2lnbl9pbl9wcm92aWRlciI6InBhc3N3b3JkIn19.iBuNAT0mvuHzoJeST36oxobZ7qc1V6JR-AK8Z8XYotPL0W8SaQwvEE3ZCUQy9XchdtgRx3eH4ZbVZArqX2s9IZb0eAZCNMwgnwXMFN722VJoYPi7KcjXJ9Y7-f_LJI-v_zby-o_u0eujV5qwUDKFJNLvuy4zW3CsYGhQEReuxRa1WVSw9HLAK1rheOweStxkIpdnStdB30YR-TAYpl97XU8weFzixA69YfvaV2hMvjaCZjqk3pa40_bQNeJSXoe6d0WTMGvXFSOPG1ceJ4yMxne8xau0xb-QIZ6b3HZayowfyFtmlpH3ClhG77Z_j_TPtXmuwtIC4A9dZSUX_A-8-w",
        "expirationTime": 1712944569589,
        "refreshToken": "AMf-vBzFX6hBa53npQqOxYLF2nGZF8ImktBebYN02wbp2FqV-0WyUuPsAn42CBYmuKKUzvN5o9WS2tVxmv9ycMB5aIjcT9U4uLULbvMMn6q0py6DIv03yrw9JEn2nHS0r4SJaxpRYrW2nEAtYnfJKahydidwt1VpXoMPJyews2OL0fMH0PXovd4CnfoyyknM9ux_r1OCXzqNfYyewztpHuNxnAAa3ii7wQ"
    },
    "tenantId": undefined,
    "uid": "XKgOMI2e6UZByzF9pM8cgt5USIb2"
}