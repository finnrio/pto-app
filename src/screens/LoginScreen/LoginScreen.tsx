import React, { useState } from "react";
import {
  Image,
  Text,
  TextInput,
  TouchableOpacity,
  View,
  Alert,
} from "react-native";
import { MaterialCommunityIcons } from '@expo/vector-icons';
import { KeyboardAwareScrollView } from "react-native-keyboard-aware-scroll-view";
import SignInWithEmailAndPassword from "../../firebase/operations/SignInWithEmailAndPassword";
import styles from "./styles";

export default function LoginScreen() {
  const [email, setEmail] = useState("");
  const [password, setPassword] = useState("");

  const HandleLogin = () => {
    SignInWithEmailAndPassword(email, password).catch((error) => {
      console.error(error.code);
      Alert.alert(error.code);
    });
  };

  return (
    <View style={styles.container}>
      <KeyboardAwareScrollView
        style={{ flex: 1, width: "100%" }}
        keyboardShouldPersistTaps="always"
      >
        <MaterialCommunityIcons name="timetable" size={120} style={styles.logo} />
        <TextInput
          style={styles.input}
          placeholder="E-mail"
          placeholderTextColor="#aaaaaa"
          onChangeText={(text) => {
            setEmail(text);
          }}
          value={email}
          underlineColorAndroid="transparent"
          autoCapitalize="none"
          testID="email_input"
        />
        <TextInput
          style={styles.input}
          placeholderTextColor="#aaaaaa"
          secureTextEntry
          placeholder="Password"
          onChangeText={(text) => {
            setPassword(text);
          }}
          value={password}
          underlineColorAndroid="transparent"
          autoCapitalize="none"
          testID="password_input"
        />
        <TouchableOpacity
          style={styles.button}
          onPress={() => {
            HandleLogin();
          }}
          testID="login_button"
        >
          <Text style={styles.buttonTitle}>Log in</Text>
        </TouchableOpacity>
      </KeyboardAwareScrollView>
    </View>
  );
}
