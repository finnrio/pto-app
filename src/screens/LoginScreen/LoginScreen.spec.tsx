import React from "react";
import { Alert } from "react-native";
import {
  cleanup,
  fireEvent,
  render,
  screen,
} from "@testing-library/react-native";
import SignInWithEmailAndPassword from "../../firebase/operations/SignInWithEmailAndPassword";
import LoginScreen from "./LoginScreen";

// solution from https://github.com/APSL/react-native-keyboard-aware-scroll-view/issues/493#issuecomment-1023551697
jest.mock("react-native-keyboard-aware-scroll-view", () => ({
  KeyboardAwareScrollView: (props: { children: React.ReactNode }) =>
    props.children,
}));

jest.mock("../../firebase/operations/SignInWithEmailAndPassword", () => {
  return jest.fn().mockImplementation(() => {
    return Promise.resolve();
  });
});

const spyAlert = jest.spyOn(Alert, "alert");
describe("LoginScreen", () => {
  afterEach(() => {
    cleanup();
    jest.clearAllMocks();
  });

  it("renders correctly", () => {
    const { toJSON } = render(<LoginScreen />);
    expect(toJSON()).toMatchSnapshot();
  });

  it("should render email and password input fields", () => {
    render(<LoginScreen />);
    expect(screen.getByTestId("email_input")).toBeTruthy();
    expect(screen.getByTestId("password_input")).toBeTruthy();
  });

  it("should call signInWithEmailAndPassword method when login button is pressed with provided email and password", async () => {
    render(<LoginScreen />);
    const email = "test@test.test";
    const password = "Password123";
    fireEvent.changeText(screen.getByTestId("email_input"), email);
    fireEvent.changeText(screen.getByTestId("password_input"), password);
    fireEvent.press(screen.getByTestId("login_button"));
    expect(SignInWithEmailAndPassword).toHaveBeenCalledWith(email, password);
  });

  it("should display an alert if an error occurs during sign in", async () => {
    require("../../firebase/operations/SignInWithEmailAndPassword").mockRejectedValueOnce(
      {
        code: "auth/invalid-email",
      },
    );
    render(<LoginScreen />);
    await fireEvent.press(screen.getByTestId("login_button"));
    expect(spyAlert).toHaveBeenCalledWith("auth/invalid-email");
  });
});
