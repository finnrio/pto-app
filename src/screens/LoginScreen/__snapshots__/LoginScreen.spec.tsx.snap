// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginScreen renders correctly 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "flex": 1,
    }
  }
>
  <Text />
  <TextInput
    autoCapitalize="none"
    onChangeText={[Function]}
    placeholder="E-mail"
    placeholderTextColor="#aaaaaa"
    style={
      {
        "backgroundColor": "white",
        "borderRadius": 5,
        "height": 48,
        "marginBottom": 10,
        "marginLeft": 30,
        "marginRight": 30,
        "marginTop": 10,
        "overflow": "hidden",
        "paddingLeft": 16,
      }
    }
    testID="email_input"
    underlineColorAndroid="transparent"
    value=""
  />
  <TextInput
    autoCapitalize="none"
    onChangeText={[Function]}
    placeholder="Password"
    placeholderTextColor="#aaaaaa"
    secureTextEntry={true}
    style={
      {
        "backgroundColor": "white",
        "borderRadius": 5,
        "height": 48,
        "marginBottom": 10,
        "marginLeft": 30,
        "marginRight": 30,
        "marginTop": 10,
        "overflow": "hidden",
        "paddingLeft": 16,
      }
    }
    testID="password_input"
    underlineColorAndroid="transparent"
    value=""
  />
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#788eec",
        "borderRadius": 5,
        "height": 48,
        "justifyContent": "center",
        "marginLeft": 30,
        "marginRight": 30,
        "marginTop": 20,
        "opacity": 1,
      }
    }
    testID="login_button"
  >
    <Text
      style={
        {
          "color": "white",
          "fontSize": 16,
          "fontWeight": "bold",
        }
      }
    >
      Log in
    </Text>
  </View>
</View>
`;
